src/util/settings/shortcuts.cc

```diff
@@ -0,0 +1,473 @@
+  return parentItem->children.size();
+}
+
+int ShortcutsModel::columnCount(const QModelIndex& parent) const {
+  // Action name and shortcuts
+  return 2;
+}
+
+QVariant ShortcutsModel::data(const QModelIndex& index, int role) const {
+  // TODO proper icon
+  static const QIcon category_ico(":/images/open.png");
+
+  if (!index.isValid()) {
+    return QVariant();
+  }
+
+  ShortcutsItem* item = static_cast<ShortcutsItem*>(index.internalPointer());
+
+  switch(role) {
+    case Qt::DisplayRole:
+      switch(index.column()) {
+        case COLUMN_INDEX_NAME:
+          return item->displayName();
+        case COLUMN_INDEX_SHORTCUTS:
+          return item->displayShortcuts();
+        default:
+          return QVariant();
+      }
+    case Qt::ForegroundRole:
+      if (item->hasConflict()) {
+        return QColor(Qt::red);
+      }
+      return util::settings::theme::pallete().color(QPalette::Text);
+    case Qt::DecorationRole:
+      if (item->isCategory() && index.column() == COLUMN_INDEX_NAME) {
+        return category_ico;
+      } else {
+        return QVariant();
+      }
+    case CATEGORY_ROLE:
+      return item->isCategory();
+    case TYPE_ROLE:
+      return item->type();
+    case SHORTCUTS_ROLE:
+      return QVariant::fromValue(item->shortcuts());
+    default:
+      return QVariant();
+  }
+}
+
+QVariant ShortcutsModel::headerData(int section, Qt::Orientation orientation,
+                                    int role) const {
+  if (orientation != Qt::Orientation::Horizontal || role != Qt::DisplayRole) {
+    return QVariant();
+  }
+
+  switch(section) {
+    case COLUMN_INDEX_NAME:
+      return "Name";
+    case COLUMN_INDEX_SHORTCUTS:
+      return "Shortcuts";
+    default:
+      return QVariant();
+  }
+
+}
+
+ShortcutsModel* ShortcutsModel::getShortcutsModel() {
+  static ShortcutsModel* ptr = new ShortcutsModel();
+  return ptr;
+}
+
+QAction* ShortcutsModel::createQAction(ShortcutType type, QObject* parent, Qt::ShortcutContext context) {
+  return type_to_shortcut_[type]->createQAction(parent, context);
+}
+
+QAction* ShortcutsModel::createQAction(ShortcutType type, QObject* parent, QIcon icon, Qt::ShortcutContext context) {
+  return type_to_shortcut_[type]->createQAction(parent, icon, context);
+}
+
+void ShortcutsModel::addShortcut(ShortcutType type, QKeySequence& shortcut) {
+  if (shortcut.isEmpty()) {
+    return;
+  }
+  auto item = type_to_shortcut_.value(type);
+  if (!item) {
+    return;
+  }
+  if (item->addShortcut(shortcut)) {
+    if (sequence_to_shortcut_.contains(shortcut) && !sequence_to_shortcut_[shortcut].contains(item)) {
+      if (sequence_to_shortcut_[shortcut].size() > 0) {
+        if (sequence_to_shortcut_[shortcut].size() == 1) {
+          sequence_to_shortcut_[shortcut][0]->setConflict(true);
+        }
+        item->setConflict(true);
+      }
+      sequence_to_shortcut_[shortcut].append(item);
+    } else {
+      sequence_to_shortcut_[shortcut].append(item);
```

Comment:
this line is copied in both branches of the if, maybe it'd be better to extract it outside?